content/post/083-js-ipfs-0.41.md

```diff
@@ -0,0 +1,313 @@
+
+## 🌗 UnixFS v1.5
+
+Turns out, it's really important for package managers to retain file metadata, particularly last modified time (`mtime`). File `mtime` allows them to selectively sync only data that has changed. Up until now if you wanted to host a large data set on IPFS, like a package manager's repository, it would be difficult to update.
+
+"What about the permanent web?" I hear you cry. Well, this absolutely **doesn't** prevent a particular snapshot of a package manager's repository from being permanently available. Metadata just enables diffs to be imported, instead of the whole thing. So, when I say "difficult to update", like I did up there, I mean slow and/or impossible. When you have Terabytes (or more) of package data and someone publishes a new package, it's _kinda_ inconvenient to import _everything_ again, when only a little part changed. File `mtime` is a really good indicator of which things have changed, so you can use it in IPFS now! 🥳
+
+For example, there's two new options to `jsipfs add` that allow `mode` and `mtime` to be preserved as the file is added to IPFS:
+
+```console
+$ jsipfs add -r --preserve-mtime --preserve-mode ~/Desktop/gif
+added QmT6WX9McZyx5ZoisRgpsjYKDBWnYpMnBLpfAgjW5kavBA gif/yesthisisdog.jpg
+added QmXMrFfZ9zHLZKN7xP2dX76YFFhvBJsQkd4fLnTDkyR31Q gif
+```
+
+Ok, no big changes there aside from the new options, buuut, now when you list directory contents you get `Mode` and `Mtime` info:
+
+```console
+$ jsipfs ls QmXMrFfZ9zHLZKN7xP2dX76YFFhvBJsQkd4fLnTDkyR31Q -v
+Mode       Mtime                           Hash                                           Size  Name
+-rw-r--r-- Apr 16, 2018, 12:20:33 PM GMT+1 QmT6WX9McZyx5ZoisRgpsjYKDBWnYpMnBLpfAgjW5kavBA 87779 yesthisisdog.jpg
+```
+
+The coolest thing is that it's completely backwards compatible. The CID for a given file/directory only changes if you opt in to metadata, otherwise the CIDs remain the same. Hooray!
+
+There's a bunch of changes that add metadata capability to the CLI, HTTP and core API both for inputs and outputs. There's also a couple of new MFS commands `touch` and `chmod` which allow you to change the metadata whenever you like! Magic 🧙‍♂️.
+
+See the API Changes sections below for details of all the new UnixFS v1.5 stuffs.
+
+# 🏗 API Changes
+
+## Core API
+
+There are significant and breaking core API changes in this release. Please see the [migration guide](https://gist.github.com/alanshaw/04b2ddc35a6fff25c040c011ac6acf26).
+
+* IPFS is not a class that can be instantiated - use `IPFS.create`. An IPFS node instance is not an event emitter.
+* The `init` option passed to `IPFS.create` will now _not_ take any initialization steps if it is set to `false`. Previously, the repo would be initialized if it already existed. This is no longer the case. If you wish to initialize a node but only if the repo exists, pass `init: { allowNew: false }` to the constructor.
```

Comment:
Is it worth bottoming [this](https://github.com/ipfs/js-ipfs/pull/2762#discussion_r377478653) out before `0.41.0`?  If we're in agreement there will be no `.init()` command any more.

Reply:
Though I realise you are anxious to get `0.41.0` out of the door..

Reply:
Hmm, would it change this option? I think we'd still want to be able to tell the node to not init the repo if it doesn't exist wouldn't we?

Reply:
Not sure tbh.

If it were me, I'd want an IPFS node with a repo so I could use it.  If it failed to create a repo I'd want to know about it, but I can't see what I'd do with a node with no repo (e.g. via `init: false`) other than call `.init()` so I can use it for offline tasks or `.init()` then `.start()` for online ones.

As for `{ allowNew: false }` I guess maybe it'd be useful to guard against my own mis-configuration (aka 'how did that repo get there?') but it feels like an edge-case.

Reply:
Discussion continued in the PR (https://github.com/ipfs/js-ipfs/pull/2762#discussion_r377478653)